lib/systemd/user-level-services.nix

```diff
@@ -0,0 +1,88 @@
+{ nixpkgs
+, system ? "x86_64-linux"
+, pkgs ? nixpkgs.legacyPackages.${system}, ... }:
+let
+  mkVM = moduleConfig: serviceConfig: nixpkgs.lib.nixosSystem {
+    inherit system;
+    modules = [ ({ ... }: {
+
+      imports = [ moduleConfig.module ];
+
+      services.${moduleConfig.nixOsServiceName} = {
+        enable = true;
+      } // serviceConfig;
+
+    }) ];
+  };
+
+in {
+  mkActivateScript = activateScriptName:
+      { module, nixOsServiceName, primaryService, primaryPortOptionName, auxiliaryServices, oneShotServices}@moduleConfig:
+      { backupAction ? "", restoreAction ? ""}:
+      serviceConfig:
+    let
+      vm = mkVM moduleConfig serviceConfig;
+      wantedServices = [primaryService] ++ auxiliaryServices;
+      allServices =  wantedServices ++ oneShotServices;
+    in
+    pkgs.writeShellScriptBin activateScriptName ''
+      set -euo pipefail
+      export XDG_RUNTIME_DIR="/run/user/$UID"
+      mkdir -v -p "$HOME/.config/systemd/user/default.target.wants"
+
+      # Deploy may change DB schema via migration and we should be able
+      # to restore its state to a point before an update to successfully rollback
+      # failed deployment
+      echo Stopping ${primaryService}
+      systemctl --user stop ${primaryService} || true
+      echo Backing up the existing state
+      ${backupAction}
+
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        ''
+          rm -v -f -- "$HOME/.config/systemd/user/${service}"
+          ln -v -s ${vm.config.systemd.units.${service}.unit}/${service} "$HOME/.config/systemd/user/${service}"
+        ''
+        ) allServices)}
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        ''
+          rm -v -f -- "$HOME/.config/systemd/user/default.target.wants/${service}"
+          ln -v -s "$HOME/.config/systemd/user/${service}" "$HOME/.config/systemd/user/default.target.wants"
+        ''
+        ) wantedServices)}
+      systemctl --user daemon-reload
+
+      rollback() {
+        exit_code="$?"
+        if [[ $exit_code -ne 0 ]]; then
+          echo Activation failed, restoring the previous working state
+          ${restoreAction}
+        fi
+        exit "$exit_code"
+      }
+      trap rollback EXIT
+
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        ''
+          echo Restarting ${service}
+          systemctl --user restart ${service}
+        ''
+      ) auxiliaryServices)}
+
+      echo Restarting ${primaryService}
```

Comment:
This a nitpick, but looks like a duplication of the code just above. You can just apply the `map` above to `auxiliaryServices ++ [ primaryService ]`

Reply:
Done.